use c++14 for size test

lucylq opened the pull request.

mergennachin commented:
We should test size_test and transitively also libexecutorch to be compilable with C++11

If it's not possible right, let's start with C++14 and then downgrade after fixing the issues.

Next target is size_test_all_ops

huydhn commented:
@pytorchbot drci

(please ignore this, I'm testing Dr.CI)

mergennachin commented on test/build_size_test.sh:
```diff
@@ -16,8 +16,12 @@ cmake_install_executorch_lib() {
   retry cmake -DBUCK2="$BUCK2" \
+          -DCMAKE_CXX_STANDARD=14 \
+          -DCMAKE_CXX_STANDARD_REQUIRED=ON \
+          -DCMAKE_CXX_FLAGS="-Wno-gnu -Werror -Wc++17-extensions" \
```
what is this for ? specifically "-Wc++17-extensions" part.

can we do it without this flag?

lucylq replied:
Updated with comments;
- Wc++17-extensions checks that we don't use c++17 features. I think it's good to have it in the test; with -Werror, this will fail compilation if we use c++17 features.
- Wno-gnu allows us to use gnu statement expressions.

swolchok replied:
> Wno-gnu allows us to use gnu statement expressions.

What can you do with gnu statement expressions that you can't do with an immediately-invoked lambda? I haven't used them much.

lucylq replied:
It sounds like they're quite equivalent, maybe worthwhile to change them as gnu statement expressions aren't part of standard c++

mergennachin commented on test/CMakeLists.txt:
```diff
@@ -19,6 +19,9 @@
 project(size_test)
 
+# Use C++11 for size test.
+set(CMAKE_CXX_STANDARD 11)
```
why the inconsistency between this and the next file (i.e. build_size_test.sh). namely, it is setting 11 and the other one is setting 14.

lucylq replied:
I guess, there are still some changes we have to make for libexecutorch to compile with c++11. Size test binary can compile using c++11, so we can use it.

swolchok reviewed (approved):
looks fine to me

lucylq committed:
use c++14 for size test (#3441)

Summary: Pull Request resolved: https://github.com/pytorch/executorch/pull/3441

Reviewed By: swolchok

Differential Revision: D56834619

Pulled By: lucylq
